File: support/tibco71/src/examples/perl/priority.pl
#!/usr/bin/perl

#
# @(#)priority.pl	1.2
#

use Tibrv;
use Getopt::Long;

$subject1="1";
$subject2="2";

sub msgCallback
{
    my ($event, $msg, $closure) = @_;
    my ($str,$send_subj,$reply_subj);

    Tibrv::tibrvMsg_GetSendSubject($msg,$send_subj);

    Tibrv::tibrvMsg_ConvertToString($msg, $str);


    printf "Received message on subject %s: %s\n", $send_subj, $str;
}





#
# creates the internal TIB/Rendezvous machinery.
#

$status = Tibrv::tibrv_Open();
die "Tibrv::tibrv_Open failed." if $status != Tibrv::TIBRV_OK;

#
# create two queues
#
Tibrv::tibrvQueue_Create($queue1);
Tibrv::tibrvQueue_Create($queue2);

#
# set the priority
#
Tibrv::tibrvQueue_SetPriority($queue1,1);
Tibrv::tibrvQueue_SetPriority($queue2,2);

#
# Create queue group and add queues
#
Tibrv::tibrvQueueGroup_Create($group);
Tibrv::tibrvQueueGroup_Add($group,$queue1);
Tibrv::tibrvQueueGroup_Add($group,$queue2);

#
# now create listeners
#

Tibrv::tibrvEvent_CreateListener($listener1,
                                 $queue1,
                                 "msgCallback",
                                 Tibrv::TIBRV_PROCESS_TRANSPORT, 
                                 $subject1, 0);

Tibrv::tibrvEvent_CreateListener($listener1,
                                 $queue2,
                                 "msgCallback",
                                 Tibrv::TIBRV_PROCESS_TRANSPORT, 
                                 $subject2, 0);



$status = Tibrv::tibrvMsg_Create($msg);

die "Tibrv::tibrvMsg_Create failed." if $status != Tibrv::TIBRV_OK;

#
# send 10 messages on subject 1
#
$status=Tibrv::tibrvMsg_SetSendSubject($msg,$subject1);
for( $i=0 ; $i < 10; $i++ )
{
    $index=$i+1;
    $valstr="value-1-" . $index;
    Tibrv::tibrvMsg_UpdateStringEx($msg,"field",$valstr);
    Tibrv::tibrvTransport_Send(Tibrv::TIBRV_PROCESS_TRANSPORT,$msg);
}

#
# send 10 messages on subject 2
#
$status=Tibrv::tibrvMsg_SetSendSubject($msg,$subject2);
for( $i=0 ; $i < 10; $i++ )
{
    $index=$i+1;
    $valstr="value-2-" . $index;
    Tibrv::tibrvMsg_UpdateStringEx($msg,"field",$valstr);
    Tibrv::tibrvTransport_Send(Tibrv::TIBRV_PROCESS_TRANSPORT,$msg);
}

#
# Dispatch the group.  When all events are dispatched, 
# timedDispatch will time out and fall out of the loop.
#

do
{
    $status=Tibrv::tibrvQueueGroup_TimedDispatch($group,1);
}while( $status == Tibrv::TIBRV_OK );

#
# cleanup and leave.
#
Tibrv::tibrvMsg_Destroy($msg);

Tibrv::tibrv_Close();

File: support/tibco71/src/examples/c/priority.c

/*
 * Copyright (c) 1998-2002 TIBCO Software Inc.
 * All rights reserved.
 * TIB/Rendezvous is protected under US Patent No. 5,187,787.
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 *
 * @(#)priority.c	1.9
 */


/*
 * priority - demonstrating queues, groups and priorities.
 *
 * There are no parameters required to run this program.
 *
 * This sample creates two queues with different priorities, creates
 * a queue group containing those queues, then publishes messages
 * to two listeners. After messages have been published the program
 * dispatches the queue group until there are no events in the group.
 * Because queues have different priorities the callback will first
 * receive messages published into the queue with higher priority.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <signal.h>
#include <string.h>
#include <time.h>

#include "tibrv/tibrv.h"

char* subject1 = "1";
char* subject2 = "2";

/*********************************************************************/
/* Message callback class                                            */
/*********************************************************************/
void msgCallback(tibrvEvent event, tibrvMsg msg, void* closure)
{
    const char* sendSubject  = NULL;
    const char* msgString    = NULL;

    /* Get the subject name to which this message was sent */
    tibrvMsg_GetSendSubject(msg,&sendSubject);

    /* Convert the incoming message to a string */
    tibrvMsg_ConvertToString(msg,&msgString);

    printf("Received message on subject %s: %s\n",
               sendSubject, msgString);

    fflush(stdout);
}

/*********************************************************************/
/* main                                                              */
/*********************************************************************/
int main(int argc, char** argv)
{
    tibrv_status status;
    tibrvQueue queue1, queue2;
    tibrvQueueGroup group;
    tibrvEvent listener1, listener2;
    tibrvMsg msg;
    char valstr[32];
    int i;

    /* Open Tibrv */
    status = tibrv_Open();
    if (status != TIBRV_OK)
    {
        fprintf(stderr,"Error: could not open TIB/RV, status=%d, text=%s\n",
            (int)status,tibrvStatus_GetText(status));
        exit(-1);
    }

    /* Create two queues */
    tibrvQueue_Create(&queue1);
    tibrvQueue_Create(&queue2);

    /* Set priorities */
    tibrvQueue_SetPriority(queue1,1);
    tibrvQueue_SetPriority(queue2,2);

    /* Create queue group and add queues */
    tibrvQueueGroup_Create(&group);
    tibrvQueueGroup_Add(group,queue1);
    tibrvQueueGroup_Add(group,queue2);

    /* Create listeners */
    tibrvEvent_CreateListener(&listener1,
                              queue1,
                              (tibrvEventCallback)msgCallback,
                              TIBRV_PROCESS_TRANSPORT,
                              subject1,
                              NULL);
    tibrvEvent_CreateListener(&listener2,
                              queue2,
                              (tibrvEventCallback)msgCallback,
                              TIBRV_PROCESS_TRANSPORT,
                              subject2,
                              NULL);

    /* Create tibrvMsg we'll be sending */
    tibrvMsg_Create(&msg);

    /* Send 10 messages on subject1 */
    tibrvMsg_SetSendSubject(msg,subject1);
    for (i=0; i<10; i++)
    {
        sprintf(valstr,"value-1-%d",(i+1));
        tibrvMsg_UpdateString(msg,"field",valstr);
        tibrvTransport_Send(TIBRV_PROCESS_TRANSPORT,msg);
    }

    /* Send 10 messages on subject1 */
    tibrvMsg_SetSendSubject(msg,subject2);
    for (i=0; i<10; i++)
    {
        sprintf(valstr,"value-2-%d",(i+1));
        tibrvMsg_UpdateString(msg,"field",valstr);
        tibrvTransport_Send(TIBRV_PROCESS_TRANSPORT,msg);
    }

    /* Dispatch the group. When all events are
     * dispatched timedDispatch() will return
     * TIBRV_TIMEOUT so we'll break out the while() loop.
     */
    while (tibrvQueueGroup_TimedDispatch(group,1) == TIBRV_OK);

    /* Close Tibrv */
    tibrv_Close();

    return 0;
}
